File: day1/day1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

func main() {
	cwd, err := os.Getwd()
	if err != nil {
		fmt.Println(err)
	}

	data_path := filepath.Join(cwd, "day1", "data.txt")

	file, err := os.Open(data_path)
	if err != nil {
		fmt.Println(err)
	}

	scanner := bufio.NewScanner(file)
	var left [1000]int
	var right [1000]int

	var index uint = 0

	for scanner.Scan() {

		var line string = scanner.Text()
		var values []string = strings.Split(line, "   ")

		left[index], err = strconv.Atoi(values[0])
		right[index], err = strconv.Atoi(values[1])
		index++
	}
	lefts := left[:]
	rights := right[:]

	sort.Ints(lefts)
	sort.Ints(rights)
	total := 0
	for i := 0; i < 1000; i++ {
		l := lefts[i]
		r := rights[i]
		if l > r {
			total += l - r
		} else {
			total += r - l
		}
	}
	fmt.Println("Total distance:", total)
}

File: day1/day1.py
import os

DATA_PATH = os.path.join(os.getcwd(), "day1", "data.txt")


def part_one():
    a = []
    b = []
    with open(os.path.join(DATA_PATH), "r") as f:
        for line in f:
            numbers = line.split("   ")
            a.append(int(numbers[0]))
            b.append(int(numbers[1]))
    a.sort()
    b.sort()
    total = 0
    for i in range(len(a)):
        total += abs(a[i] - b[i])
    print(f"Total distance: {total}")


def part_two():
    left = []
    right = []
    right_count = {}
    with open(DATA_PATH, "r") as f:
        for line in f:
            numbers = line.split("   ")
            left.append(int(numbers[0]))

            right_number = int(numbers[1])

            if right_count.get(right_number) is None:
                right_count[right_number] = 0
            right_count[right_number] += 1

            right.append(right_number)
    sim_score = 0
    for left_num in left:
        sim_score += left_num * right_count.get(left_num, 0)
    print(f"Similarity Score: {sim_score}")


if __name__ == "__main__":
    part_one()
    part_two()
